edc-extensions/agreements/retirement-evaluation-store-sql/src/main/java/org/eclipse/tractusx/edc/agreements/retirement/store/sql/SqlAgreementsRetirementStore.java

```diff
@@ -0,0 +1,123 @@
+/********************************************************************************
+ * Copyright (c) 2024 Bayerische Motoren Werke Aktiengesellschaft (BMW AG)
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Apache License, Version 2.0 which is available at
+ * https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ ********************************************************************************/
+
+package org.eclipse.tractusx.edc.agreements.retirement.store.sql;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import org.eclipse.edc.spi.persistence.EdcPersistenceException;
+import org.eclipse.edc.spi.query.QuerySpec;
+import org.eclipse.edc.spi.result.StoreResult;
+import org.eclipse.edc.sql.QueryExecutor;
+import org.eclipse.edc.sql.store.AbstractSqlStore;
+import org.eclipse.edc.transaction.datasource.spi.DataSourceRegistry;
+import org.eclipse.edc.transaction.spi.TransactionContext;
+import org.eclipse.tractusx.edc.agreements.retirement.spi.store.AgreementsRetirementStore;
+import org.eclipse.tractusx.edc.agreements.retirement.spi.types.AgreementsRetirementEntry;
+
+import java.sql.Connection;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.Objects;
+import java.util.stream.Stream;
+
+public class SqlAgreementsRetirementStore extends AbstractSqlStore implements AgreementsRetirementStore {
+
+    private final SqlAgreementsRetirementStatements statements;
+
+    public SqlAgreementsRetirementStore(DataSourceRegistry dataSourceRegistry, String dataSourceName,
+                                        TransactionContext transactionContext, ObjectMapper objectMapper,
+                                        QueryExecutor queryExecutor, SqlAgreementsRetirementStatements statements) {
+        super(dataSourceRegistry, dataSourceName, transactionContext, objectMapper, queryExecutor);
+        this.statements = statements;
+    }
+
+    @Override
+    public StoreResult<Void> save(AgreementsRetirementEntry entry) {
+        Objects.requireNonNull(entry);
+        return transactionContext.execute(() -> {
+            try (var connection = getConnection()) {
+                if (existsById(entry.getAgreementId(), connection)) {
+                    return StoreResult.alreadyExists(ALREADY_EXISTS_TEMPLATE.formatted(entry.getAgreementId()));
+                }
+                insert(connection, entry);
```

Comment:
Because of the constraint on the `edc_contract_agreement` table, this might throw what I believe will be an uncaught exception.

Reply:
it should be handled in the next catch block
`catch (SQLException e) {
                throw new EdcPersistenceException(e);
            }`

simulated locally and confirmed
`org.eclipse.edc.spi.persistence.EdcPersistenceException: ERROR: insert or update on table "edc_agreement_retirement" violates foreign key constraint "agreement_retirement_contract_agreement_id_fk"`

Reply:
that is valid. looking into `SqlQueryExecutor.execute()` we can confirm the already handled and throw of `EdcPersistenceException` to the client, so here catching `EdcPersistenceException` to throw it once more seems redundant.

Reply:
we shouldn't rely on fk constraint, consistency should be maintained at the service layer

Reply:
I assumed that since it depends on the existence of the fk then it could be controlled here to keep consistency. 
Removed in [here](https://github.com/eclipse-tractusx/tractusx-edc/pull/1631/commits/223e47d48cb7fdc2807da6ecb9589541a30d208d).